posts/2023-10-30_floating_point/floating_point.qmd

```diff
@@ -0,0 +1,338 @@
+
+```{r}
+(101 * 1.1 * 10) %>%
+  round() %>%
+  format(digits = 22)
+
+(111.1 * 10) %>%
+  round() %>%
+  format(digits = 22)
+```
+
+This is very awkward, as you don't know by how much you need to multiply each time, a very clunky solution.
+
+Alternatively, we can compare the absolute value of the difference between the analysis value and the upper limit plus 10% to a very small number, e.g. 0.0000001:
+
+```{r}
+AVAL <- 111.1
+ANRHI <- 101
+
+abs(AVAL - ANRHI * 1.1) < 0.0000001
+```
+
+Comparing to a very small value is also how the `all.equal()` function works, which compares two numeric values and returns `TRUE` if they are equal within a tolerance.
+By default the tolerance is around $1.5 * 10^{-8}$ but you can set it yourself to a lower value, e.g. machine tolerance `.Machine$double.eps` - (one of\*\*) the smallest positive floating-point number x such that 1 + x != 1.
+
+```{r}
+1 + .Machine$double.eps == 1
+# but:
+1 + .Machine$double.eps / 2 == 1
+
+# so we can use:
+all.equal(AVAL, ANRHI * 1, 1, tolerance = .Machine$double.eps)
+```
+
+This would still be a little clunky for *greater than or equal to* comparisons: 
+
+```{r}
+all.equal(AVAL, ANRHI * 1.1) | AVAL > ANRHI * 1.1
+# unfortunately, the all.equal() function does not return a FALSE if they are not the same:
+all.equal(AVAL, ANRHI * 1.1 + 1)
+```
+
+For some reason, the value it returns is also not correct.
+
+There is also a dplyr function called `near()` which does essentially the same thing as `all.equal()`:
+
+```{r}
+ANRHI <- 100
+AVAL <- 110
+(ANRHI * 1.1) %>% format(digits = 22)
+AVAL > ANRHI * 1.1 | near(AVAL, ANRHI * 1.1)
+```
+
+By Gordon Miller suggested to replace the standard comparators with the following functions across [{{admiral}}](https://github.com/pharmaverse/admiral)
+
+| `base` | improved             |
+|----------|----------------------|
+| A \>= B  | A \> B \| near(A, B) |
+| A \<= B  | A \< B \| near(A, B) |
+| A == B   | near(A, B)           |
+| A != B   | !near(A, B)           |
+| A \> B   | A \> B & !near(A, B) |
+| A \< B   | A \< B & !near(A, B) |
+
+
+This would work perfectly fine, but especially for `case_when()` statements, it would add a lot of code-bloat.
+
+
+::: callout-note
+Although a minor issue, it looks like the `near()` function tests for absolute differences, while the `all.equal()` function tests for relative differences, as discussed in [this thread](https://github.com/tidyverse/dplyr/issues/6921):
+
+```{r}
+# very large values:
+# when checking for absolute differences
+near(
+  ANRHI * 1.1 * 10^6,
+  AVAL * 10^6
+)
+# when checking for relative differences
+all.equal(
+  ANRHI * 1.1 * 10^6,
+  AVAL * 10^6
+)
+
+# as:
+(ANRHI * 1.1 * 10^6) %>% format(digits = 22)
+(AVAL * 10^6) %>% format(digits = 22)
+```
+
+:::
+
+A more elegant solution is implemented in the `fpCompare` package, which compares floating point numbers using custom infix operators: 
```

Comment:
```suggestion
```{r}
# Very large values:
# When checking for absolute differences
near(
  ANRHI * 1.1 * 10^6,
  AVAL * 10^6
)
# When checking for relative differences
all.equal(
  ANRHI * 1.1 * 10^6,
  AVAL * 10^6
)

# As:
(ANRHI * 1.1 * 10^6) %>% format(digits = 22)
(AVAL * 10^6) %>% format(digits = 22)
```

:::

A more elegant solution is implemented in the `fpCompare` package, which compares floating point numbers using custom infix operators: 
```